src/main/java/org/broadinstitute/barclay/argparser/TaggedArgumentParser.java

```diff
@@ -0,0 +1,333 @@
+ * name, and tag string), and the raw argument value together to ensure the key is unique:
+ *
+ *      Key: --argument_name:logical_name,key1=value1,key2=value2:argument_value
+ *
+ * The pair object stored in the map contains only the (previously peeled off) tag string, and the raw argument value that
+ * was provided by the user. The command line parser replaces the original arguments provided by the user with the option
+ * name and the key:
+ *
+ *          User provides:              "--argument_name:logical_name,key1=value1,key2=value2" "raw_argument_value"
+ *          Parser is presented with:   "--argument_name" "--argument_name:logical_name,key1=value1,key2=value2:argument_value"
+ *          Map stores:                 Pair("logical_name,key1=value1,key2=value2", "argument_value")
+ *
+ * 2) In phase 2, which occurs when the underlying argument field is being populated with a value, the key is used to
+ * retrieve the original tag string and argument value. The tag string is parsed (logical name and attributes) and used
+ * to populate the underlying argument field.
+ */
+public final class TaggedArgumentParser {
+
+    /**
+     * Delimiter between key-value pairs in the "logical_name,key1=value1,key2=value2" syntax.
+     */
+    private static final String ARGUMENT_KEY_VALUE_PAIR_DELIMITER = ",";
+
+    /**
+     * Separator between keys and values in the "logical_name,key1=value1,key2=value2" syntax.
+     */
+    private static final String ARGUMENT_KEY_VALUE_SEPARATOR = "=";
+
+    /**
+     * Separator used between option name and logical name.
+     */
+    private static final char ARGUMENT_TAG_NAME_SEPARATOR = ':';
+
+    private static final String USAGE = "Argument must of the form argument_name:logical_name(,key=value)*";
+
+    // Map of surrogate keys to Pair(tag_string, raw_argument_value)
+    private Map<String, Pair<String, String>> tagSurrogates = new HashMap<>();
+
+    // Prefixes used by the opt parser for short/long prefixes
+    private static final String SHORT_OPTION_PREFIX = "-";
+    private static final String LONG_OPTION_PREFIX = "--";
+
+    /**
+     * Given an array of raw arguments provided by the user, return an array of args where tagged arguments
+     * have been replaced with curated arguments containing a key to be used by the parser to retrieve the actual
+     * values.
+     * @param args raw arguments as provided by the user
+     * @return curated string of arguments to be presented to the opt parser
+     */
+    public String[] processTaggedOptions(final String[] args) {
+        List<String> finalArgs = new ArrayList<>(args.length);
+        for (int i = 0; i < args.length; ) {
+            String arg = args[i];
+            if (isShortOptionToken(arg)) {
+                i += replaceTaggedOption(SHORT_OPTION_PREFIX, arg.substring(SHORT_OPTION_PREFIX.length()), args, finalArgs, i);
```

Comment:
Manipulating a loop counter within a for loop is usually a bad idea/error prone -- I think you could avoid the need for this by converting from an array to an iterator of Strings.